### PR TITLE
Updating unimarshaller

### PR DESCRIPTION
**Pull request by mfdeveloper508**
https://github.com/moov-io/iso8583/issues/282
https://github.com/moov-io/iso8583/issues/279

### CODE REVIEW

File: field/binary.go [hunk 1]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
 }
 
 func (f *Binary) Unmarshal(v interface{}) error {
-	if v == nil {
-		return nil
+	switch val := v.(type) {
+	case reflect.Value:
+		switch val.Kind() { //nolint:exhaustive
```
Comment (alovak) [inline field/binary.go:122]:
```suggestion
	case reflect.Value:
		if !val.CanSet() {
			return fmt.Errorf("cannot set reflect.Value of type %s", val.Kind())
		}

		switch val.Kind() { //nolint:exhaustive
```

File: field/binary.go [hunk 2]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+		case reflect.String:
+			if !val.CanSet() {
+				return fmt.Errorf("reflect.Value of the data can not be change")
+			}
```
Comment (alovak) [inline field/binary.go:126]:
```suggestion
```

File: field/binary.go [hunk 3]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+
+			str := hex.EncodeToString(f.value)
+			val.SetString(str)
+		case reflect.Slice:
+			if !val.CanSet() {
+				return fmt.Errorf("reflect.Value of the data can not be change")
+			}
+
```
Comment (alovak) [inline field/binary.go:134]:
```suggestion
```

File: field/binary.go [hunk 4]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+			val.SetBytes(f.value)
+		default:
+			return fmt.Errorf("data does not match required reflect.Value type")
+		}
+	case *string:
+		str := hex.EncodeToString(f.value)
+		*val = str
```
Comment (alovak) [inline field/binary.go:141]:
```suggestion
		*val = hex.EncodeToString(f.value)
```

File: field/binary.go [hunk 5]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+			str := hex.EncodeToString(f.value)
+			val.SetString(str)
+		case reflect.Slice:
+			if !val.CanSet() {
+				return fmt.Errorf("reflect.Value of the data can not be change")
+			}
+
+			val.SetBytes(f.value)
+		default:
+			return fmt.Errorf("data does not match required reflect.Value type")
```
Comment (alovak) [inline field/binary.go:137]:
```suggestion
			return fmt.Errorf("unsupported reflect.Value type: %s", val.Kind())
```

File: field/binary.go [hunk 6]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+		}
+	case *string:
+		str := hex.EncodeToString(f.value)
+		*val = str
+	case *[]byte:
+		*val = f.value
+	case *Binary:
+		val.value = f.value
+	default:
+		return fmt.Errorf("data does not match required *Binary or (*string, *[]byte) type")
```
Comment (alovak) [inline field/binary.go:147]:
```suggestion
		return fmt.Errorf("unsupported type: expected *Binary, *string, *[]byte, or reflect.Value, got %T", v)
```

File: field/binary.go [hunk 7]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+		}
+
+		buf, err := hex.DecodeString(*v)
+		if err != nil {
+			return fmt.Errorf("failed to convert string to byte: %w", err)
+		}
+
+		f.value = buf
+	case []byte:
+		f.SetBytes(v)
```
Comment (alovak) [inline field/binary.go:185]: It can also be `nil`.

File: field/binary.go [hunk 8]
```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
...
+	case *[]byte:
+		if v == nil {
+			f.value = nil
+			return nil
+		}
+		f.SetBytes(*v)
+	default:
+		return fmt.Errorf("data does not match required *Binary or (string, *string, []byte, *[]byte) type")
```
Comment (alovak) [inline field/binary.go:196]: @mfdeveloper508 I suggest us to rework the method a little bit. First, we always set f.value to nil if `v` has zero value and return. So we can do this only once and the rest of the code will not have `if` statements. It seems the following code should work:


```go
func (f *Binary) Marshal(v interface{}) error {
	if reflect.ValueOf(v).IsZero() {
		f.value = nil
		return nil
	}

	switch val := v.(type) {
	case *Binary:
		f.value = val.value
	case string:
		buf, err := hex.DecodeString(val)
		if err != nil {
			return utils.NewSafeError(fmt.Errorf("converting string (as hex) to byte: %w", err), "failed to convert string (as hex) to byte") 
		}
		f.value = buf
	case *string:
		buf, err := hex.DecodeString(*val)
		if err != nil {
			return utils.NewSafeError(fmt.Errorf("converting string (as hex) to byte: %w", err), "failed to convert string (as hex) to byte") 
		}
		f.value = buf
	case []byte:
		f.SetBytes(val)
	case *[]byte:
		f.SetBytes(*val)
	default:
		return fmt.Errorf("data does not match required *Binary or (string, *string, []byte, *[]byte) type")
	}

	return nil
}
```

File: field/binary_test.go [hunk 9]
```diff
@@ -140,3 +141,76 @@ func TestBinaryNil(t *testing.T) {
...
+	err = str.Unmarshal(val5)
+	require.NoError(t, err)
+	require.Equal(t, testValue, val5.Bytes())
+
+	val6 := reflect.ValueOf(val2)
+	err = str.Unmarshal(val6)
+	require.Error(t, err)
+	require.Equal(t, "cannot set reflect.Value of type string", err.Error())
+
+	val7 := reflect.ValueOf(&val2)
```
Comment (alovak) [inline field/binary_test.go:179]: I would suggest to name variables a little bit differently. In this line, it's not clear what is the type of `val2`.

What about the following changes:

```
testValue := []byte{0x12, 0x34, 0x56} -> expectedBytes
val1 := &Binary{} -> binary or binaryField
var val2 string -> str or s
val3 []byte -> bytes
```

If you agree, we can do this for all tests.

Comment (mfdeveloper508) [reply field/binary_test.go:179]: agree

**Review by alovak — APPROVED** [review]
Great job!

**Pull request merged**